integration/run_test.go

```diff
@@ -0,0 +1,128 @@
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package integration
+
+import (
+	"fmt"
+	"regexp"
+	"testing"
+	"time"
+
+	"github.com/awslabs/soci-snapshotter/util/testutil"
+)
+
+// TestRunMultipleContainersWithSameImage runs two containers with the same image at the same time
+func TestRunMultipleContainersWithSameImage(t *testing.T) {
+	RunMultipleWebContainers(t, []string{nginxImage, nginxImage})
+}
+
+// TestRunMultipleContainersWithDifferentImages runs two containers with different images at the same time
+func TestRunMultipleContainersWithDifferentImages(t *testing.T) {
+	RunMultipleWebContainers(t, []string{nginxImage, drupalImage})
+}
+
+func RunMultipleWebContainers(t *testing.T, imageNames []string) {
+	regConfig := newRegistryConfig()
+	sh, done := newShellWithRegistry(t, regConfig)
+	defer done()
+
+	// Initialize config files for containerd and snapshotter
+	additionalConfig := ""
+	if !isTestingBuiltinSnapshotter() {
+		additionalConfig = proxySnapshotterConfig
+	}
+	containerdConfigYaml := testutil.ApplyTextTemplate(t, `
+version = 2
+
+[plugins."io.containerd.snapshotter.v1.soci"]
+root_path = "/var/lib/soci-snapshotter-grpc/"
+
+[plugins."io.containerd.snapshotter.v1.soci".blob]
+check_always = true
+
+{{.AdditionalConfig}}
+`, struct {
+		AdditionalConfig string
+	}{
+		AdditionalConfig: additionalConfig,
+	})
+
+	// Setup environment
+	if err := testutil.WriteFileContents(sh, defaultContainerdConfigPath, []byte(containerdConfigYaml), 0600); err != nil {
+		t.Fatalf("failed to write %v: %v", defaultContainerdConfigPath, err)
+	}
+	sh.
+		X("update-ca-certificates").
+		Retry(100, "nerdctl", "login", "-u", regConfig.user, "-p", regConfig.pass, regConfig.host)
+
+	rebootContainerd(t, sh, "", "")
+	for _, imageName := range deduplicate(imageNames) {
+		// Mirror image
+		copyImage(sh, dockerhub(imageName), regConfig.mirror(imageName))
+		indexDigest := optimizeImage(sh, regConfig.mirror(imageName))
+
+		// Pull image
+		sh.X("ctr", "i", "pull", "--user", regConfig.creds(), regConfig.mirror(imageName).ref)
+
+		// Create SOCI metadata
+		sh.X("soci", "create", regConfig.mirror(imageName).ref)
```

Comment:
The `optimizeImage` function already creates a soci index.

Reply:
Good catch, thanks. Can you confirm that https://github.com/awslabs/soci-snapshotter/blob/main/integration/pull_test.go#L786-L794 is also redundant?

Reply:
Hmm, looks like it. Should probably reach out to @vkuzniet about this; he will have the most context.

Reply:
Yep, it is redundant. Pull and create happen during `optimizeImage`.

Reply:
I'm fixing this in my PR. I'll probably open another PR later to clean up the redundancy elsewhere.